packages/langium/src/parser/parser-builder-base.ts

```diff
@@ -0,0 +1,419 @@
+        }
+        return ruleArgs;
+    };
+}
+
+interface PredicatedMethod {
+    ALT: Method,
+    GATE?: Predicate
+}
+
+function buildPredicate(condition: Condition): Predicate {
+    if (isDisjunction(condition)) {
+        const left = buildPredicate(condition.left);
+        const right = buildPredicate(condition.right);
+        return (args) => (left(args) || right(args));
+    } else if (isConjunction(condition)) {
+        const left = buildPredicate(condition.left);
+        const right = buildPredicate(condition.right);
+        return (args) => (left(args) && right(args));
+    } else if (isNegation(condition)) {
+        const value = buildPredicate(condition.value);
+        return (args) => !value(args);
+    } else if (isParameterReference(condition)) {
+        const name = condition.parameter.ref!.name;
+        return (args) => args !== undefined && args[name] === true;
+    } else if (isLiteralCondition(condition)) {
+        const value = !!condition.true;
+        return () => value;
+    }
+    assertUnreachable(condition);
+}
+
+function buildAlternatives(ctx: RuleContext, alternatives: Alternatives): Method {
+    if (alternatives.elements.length === 1) {
+        return buildElement(ctx, alternatives.elements[0]);
+    } else {
+        const methods: PredicatedMethod[] = [];
+
+        for (const element of alternatives.elements) {
+            const predicatedMethod: PredicatedMethod = {
+                // Since we handle the guard condition in the alternative already
+                // We can ignore the group guard condition inside
+                ALT: buildElement(ctx, element, true)
+            };
+            const guard = getGuardCondition(element);
+            if (guard) {
+                predicatedMethod.GATE = buildPredicate(guard);
+            }
+            methods.push(predicatedMethod);
+        }
+
+        const idx = ctx.or++;
+        return (args) => ctx.parser.alternatives(idx, methods.map(method => {
+            const alt: IOrAlt<unknown> = {
+                ALT: () => method.ALT(args)
+            };
+            const gate = method.GATE;
+            if (gate) {
+                alt.GATE = () => gate(args);
+            }
+            return alt;
+        }));
+    }
+}
+
+function buildUnorderedGroup(ctx: RuleContext, group: UnorderedGroup): Method {
+    if (group.elements.length === 1) {
+        return buildElement(ctx, group.elements[0]);
+    }
+    const methods: PredicatedMethod[] = [];
+
+    for (const element of group.elements) {
+        const predicatedMethod: PredicatedMethod = {
+            // Since we handle the guard condition in the alternative already
+            // We can ignore the group guard condition inside
+            ALT: buildElement(ctx, element, true)
+        };
+        const guard = getGuardCondition(element);
+        if (guard) {
+            predicatedMethod.GATE = buildPredicate(guard);
+        }
+        methods.push(predicatedMethod);
+    }
```

Comment:
The for loop above is almost identical to the one in `buildAlternatives`. Perhaps these could be factored out into a common function `buildMethods` that takes the necessary elements and context?